### PR TITLE
Fixed upper and mixed case columns not quoted.

### PR DESCRIPTION
**Pull request by paulrrogers**
|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Fixed issues | #3710

#### Summary

Checks for presence of upper and lowercase letters and if so then quotes the value

### CODE REVIEW

**Comment by morozov** [comment]
@paulrrogers than you for the pull request. Since it affects a platform-specific behavior, I'd prefer to have a functional test that would guarantee that the new code works with the target platform. Not just theory by making assertions against an implementation detail.

**Comment by paulrrogers** [comment]
> @paulrrogers thank you for the pull request. Since it affects a platform-specific behavior, I'd prefer to have a functional test that would guarantee that the new code works with the target platform. Not just in theory by making assertions against an implementation detail.

Sorry if I'm not up-to-speed on the terminology here. By 'functional' do you mean a test that covers the integration with the platform? I.e. it covers running queries through a PostgreSQL instance?

**Comment by morozov** [comment]
> By 'functional' do you mean a test that covers the integration with the platform? I.e. it covers running queries through a PostgreSQL instance?

Yes.

**Comment by paulrrogers** [comment]
@morozov I've added a functional test, though not sure why portability tests like `testFetchAllNullColumn` are now failing. Perhaps Doctrine is doing some mocking that has been disrupted or my fix is incomplete?

Can you offer any insight?

**Comment by morozov** [comment]
Not sure why it's failing but the currently provided test is not what I had in mind. Please look at the [`DateExpressionTest`](https://github.com/doctrine/dbal/blob/v2.9.2/tests/Doctrine/Tests/DBAL/Functional/Platform/DateExpressionTest.php) for example. It tests the code that implements a database abstraction by interacting with the real database. That proves that the code works correctly, not just as expected.

I didn't look at the related issue but from the test itself, it's not clear what problem the patch solves. It should be.

**Comment by paulrrogers** [comment]
> …the currently provided test is not what I had in mind. Please look at the [`DateExpressionTest`](https://github.com/doctrine/dbal/blob/v2.9.2/tests/Doctrine/Tests/DBAL/Functional/Platform/DateExpressionTest.php) for example. It tests the code that implements a database abstraction by interacting with the real database. That proves that the code works correctly, not just as expected.

My apologies, didn't realize I forgot to check-in `testQuotingMixedCaseAssets`. Please let me know if that is clear enough.

File: lib/Doctrine/DBAL/Schema/AbstractAsset.php [hunk 1]
```diff
@@ -177,10 +178,12 @@ public function getName()
...
     public function getQuotedName(AbstractPlatform $platform)
     {
-        $keywords = $platform->getReservedKeywordsList();
-        $parts    = explode('.', $this->getName());
+        $keywords     = $platform->getReservedKeywordsList();
+        $isCaseFolded = $platform->isCaseFoldedForUnquotedAssets();
+        $parts        = explode('.', $this->getName());
         foreach ($parts as $k => $v) {
-            $parts[$k] = $this->_quoted || $keywords->isKeyword($v) ? $platform->quoteIdentifier($v) : $v;
+            $caseNeedsQuoting = $isCaseFolded && preg_match('/(\p{Ll}.*\p{Lu}|\p{Lu}.*\p{Ll})/u', $v);
```
Comment (SenseException) [inline lib/Doctrine/DBAL/Schema/AbstractAsset.php]: I'd like to see this line in a separate private method. This way, the regex might not need to run when the first two matches are true in the next line.

**Pull request closed without merging**